### ISSUE TITLE
Dual media element playback should be specified for ad insertion, etc.

### ISSUE THREAD

**Issue by KilroyHughes** [w3c/webmediaporting#27]
[Kilroy action agreed at Amsterdam F2F]
It is common practice today to use separate media elements for the primary program and interstitial video ads, often using a plugin, such as Flash. WAVE should specify recommendations for HTML5 player applications and platforms to use and support dual media elements. 

Background:
A separate media element for ads allows them to use different media profiles, encryption, Track options, delivery method, and latency from the main program, without interruption and delay caused by resetting the main media pipeline. Presentation focus can be switched between media elements on any frame.

The alternative of splice conditioning ad and main program content may not be possible for use cases such as personalized ad insertion by each player, ad networks where each ad may be inserted into thousands of different programs and services determined at the moment of playback, and live programs where segmentation of the live stream can't be personalized for the frame accurate duration of different personalized ad pods. A single media element also prevents low latency live streaming if the ad network has higher latency.  A separate ad element and track buffer allows ads to be delivered and queued any time in advance of playback. Different picture aspect ratios, codecs, and frame rates can be better accommodated. A simple file source and file download can be used for short ads, multiplexed M2TS ads can overlay CMAF programs, and unencrypted ads can be visually interspersed with encrypted content running in a protected video path. 

In the case of an embedded (Type 1) MPD or M3U8 player, playback of the main program can continue independently from ads and GUI rendered in a media element in a web page or native app, so the ad insertion app and media element function the same whether it is in a Type 1 or Type 3 environment. 

There are many reasons why ad networks use separate ad insertion plugins across various types of streaming players and media formats, separate from the delivery and decoding advantages of a separate media element. They are used for player authentication and service authorization (a plugin can test and verify different players are trusted ad players), personalized and secure identification and target metadata in ad requests, cryptographically authenticated and protected ad progress and demographic reporting to the ad network's measurement server, enforcement of ad skipping rules in the main player, user interaction with ads, etc. 

Using a single media element and a continuous stream of splice conditioned video segments works well when ad and program encoding are controlled by a single entity for a single program. The same codecs and encoding parameters can be used (sample entries, indexes, etc.) and live program CVSs re-encoded, time stamped, and segmented to match the end of each ad pod. Targeted ads must be splice conditioned for exact substitution via playlist/manifest, or the same main program delivered and cached as different programs if they have different ads, which is much less efficient. Programs that are not splice conditioned often result in some delay at each splice point as the media pipeline, DRM, etc. are reconfigured; resulting in an accumulating presentation delay in live programs.

**Comment by jpiesing** [comment]
I think the title of this is a little misleading based on the problem description. The description implies two complete media pipelines and not just two media elements. HbbTV requires support for multiple media elements but only a single media pipeline which can be switched from one media element to another in defined circumstances and within a defined period if certain conditions are met. I assume the HbbTV approach isn't what is meant here.

**Comment by KilroyHughes** [comment]
Yes, the title only mentions the HTML5 API aspect, but the requirements for apps and user agents also need to be defined.

I didn’t want to jump to the solution without discussion, but it is clear from the problem statement that separate media pipelines are needed to avoid the problems caused by tearing down and rebuilding a single media pipeline between the last video frame of one CMAF Presentation (e.g. an ad) and the first frame of an unrelated CMAF Presentation (e.g. that instant in a live stream).

Plugins like Flash and Silverlight have typically used their own software decoders. Browsers often contain their own software decoders, and have different policies on defaulting to hardware decoders and DRMs when available at the OS/hardware layer. A shared media pipeline canvas and scene graph/DOM isn’t a problem. A separate software decoder is probably necessary for ads, while the main program maintains exclusive use of platform embedded DRM and hardware decoding (possibly a complete Type 1 player).

These are some of the fundamental issues we need to resolve as we move from web media based on plugins, to fully functional (and secure) native HTML5 media APIs.

Kilroy Hughes

**Comment by jpiesing** [comment]
> Yes, the title only mentions the HTML5 API aspect, but the requirements for apps and user agents also need to be defined. I didn’t want to jump to the solution without discussion, but it is clear from the problem statement that separate media pipelines are needed to avoid the problems caused by tearing down and rebuilding a single media pipeline between the last video frame of one CMAF Presentation (e.g. an ad) and the first frame of an unrelated CMAF Presentation (e.g. that instant in a live stream).

My comment is just that the title of the issue is misleading as there are existing implementations of the title that don't deliver what is in the problem statement. IMHO the title should reflect the first paragraph of the background. Perhaps something like "seamless switching between content and ads without splice conditioning"?
